staging/src/k8s.io/apiserver/pkg/apis/apiserverinternal/v1alpha1/types.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+//  Storage version of a specific resource.
+type StorageVersion struct {
+	metav1.TypeMeta `json:",inline"`
+	// The name is <group>.<resource>.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec is an empty spec. It is here to comply with Kubernetes API style.
+	Spec StorageVersionSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// API server instances report the version they can decode and the version they
+	// encode objects to when persisting objects in the backend.
+	Status StorageVersionStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// StorageVersionSpec is an empty spec.
+type StorageVersionSpec struct{}
+
+// API server instances report the versions they can decode and the version they
+// encode objects to when persisting objects in the backend.
+type StorageVersionStatus struct {
+	// The reported versions per API server instance.
+	// +optional
+	// +listType=map
+	// +listMapKey=apiserverID
+	ServerStorageVersions []ServerStorageVersion `json:"serverStorageVersions,omitempty" protobuf:"bytes,1,opt,name=serverStorageVersions"`
+	// If all API server instances agree on the same encoding storage version,
+	// then this field is set to that version. Otherwise this field is left empty.
+	// +optional
+	AgreedEncodingVersion *string `json:"agreedEncodingVersion,omitempty" protobuf:"bytes,2,opt,name=agreedEncodingVersion"`
+
+	// The latest available observations of the storageVersion's state.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []StorageVersionCondition `json:"conditions,omitempty" protobuf:"bytes,3,opt,name=conditions"`
+}
+
+// An API server instance reports the version it can decode and the version it
+// encodes objects to when persisting objects in the backend.
+type ServerStorageVersion struct {
+	// The ID of the reporting API server.
+	// For a kube-apiserver, the ID is configured via a flag.
```

Comment:
Hm, if we're writing endpoints, we write something. Can we use that? Not super thrilled about administrators having to remember to to set a flag differently.

If we're worried about letting some detail cross from a hidden network, I can imagine e.g. salt+hash hostname and/or IP address to produce this.

Reply:
> Can we use that? 
No. When writing to endpoints, apiserver uses the value of the `--advertise-address` flag. I know for regional clusters in GKE, the flag is set to the IP of loadbalancer, so all apiservers appear the same. Possibly other providers do the same.

Maybe we can use hash(salt+$advertise-address), but since that's going to change after every reboot anyway, then why don't we just use a random UID?

Anyway, this is beyond the API definition. Can we iterate on this later?

Reply:
OK sure. Random is probably fine.

Reply:
I removed it from the comment. I can add it back when we actually implement the ID.